morango/syncsession.py

```diff
@@ -220,56 +270,154 @@ class SyncClient(object):
+
+    def _queuing(self, data, push):
+        if push:
+            self._queue_into_buffer()
+            # update the records_total for client and server transfer session
+            records_total = Buffer.objects.filter(transfer_session=self.current_transfer_session).count()
+            self.current_transfer_session.records_total = records_total
+            self.current_transfer_session.transfer_stage = transfer_status.PUSHING
+            self.current_transfer_session.save()
+        else:
+            # creating transfer session on pull also queues data server side
+            try:
+                response = self.sync_connection._create_transfer_session(data)
+            except requests.HTTPError as e:
+                self.current_transfer_session.delete(soft=True)
+                raise
+
+            self.current_transfer_session.server_fsic = response.json().get('server_fsic') or '{}'
+            self.current_transfer_session.records_total = response.json().get('records_total')
+            self.current_transfer_session.transfer_stage = transfer_status.PULLING
+            self.current_transfer_session.save()
 
-    def initiate_push(self, sync_filter):
+    def _push_records(self, callback=None):
+        buffered_records = Buffer.objects.filter(transfer_session=self.current_transfer_session).order_by('pk')
 
-        self._create_transfer_session(True, sync_filter)
+        while self.current_transfer_session.records_transferred < self.current_transfer_session.records_total:
+            # slice queryset to get chunk of records for sending
+            records = buffered_records[self.current_transfer_session.records_transferred:self.chunk_size + self.current_transfer_session.records_transferred]
+            # serialize and send records to server
+            serialized_recs = BufferSerializer(records, many=True)
+            logger.info('Pushing {} records at a time with {}/{} records transferred'.format(self.chunk_size,
+                                                                                             self.current_transfer_session.records_transferred,
+                                                                                             self.current_transfer_session.records_total))
+            try:
+                self.sync_connection._push_record_chunk(serialized_recs.data)
+            except requests.HTTPError as e:
+                self._close_transfer_session()
+                raise
 
-        _queue_into_buffer(self.current_transfer_session)
+            # update records_transferred upon successful request
+            self.current_transfer_session.records_transferred += self.chunk_size
+            # ensure records transferred does not go over total records
+            if self.current_transfer_session.records_transferred > self.current_transfer_session.records_total:
+                self.current_transfer_session.records_transferred = self.current_transfer_session.records_total
```

Comment:
Would it be safer (if perhaps slightly slower) to just add `len(records)` here, instead of `self.chunk_size`? Then it would never go over, so this could be an `assert` instead of a tweak?

Reply:
I could do `records.count()`